### PR TITLE
Safer startup logic

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Don't use snapctl on `run-daemon`, it is racy and can fail some of the time on some systems.

See https://forum.snapcraft.io/t/scummvm-snap-failing-to-install-on-rpi-4/21394/59

### CODE REVIEW

**Review by wmww — APPROVED** [review]
Looks good to me. *if* you wanted me to test that it works I'd need more context for what it's supposed to be doing in what situations.

**Comment by AlanGriffiths** [comment]
> Looks good to me. _if_ you wanted me to test that it works I'd need more context for what it's supposed to be doing in what situations.

1. On installation `daemon` should default to "true" on Ubuntu Core systems and "false" on "classic" systems;
2. The `mir-kiosk.daemon` daemon should run if and only if `daemon=true`; and,
3. (what this changes) `snapctrl stop ...` should not be called from the daemon launch script, only from hooks.

File: snap/hooks/install [hunk 1]
```diff
@@ -1,6 +1,10 @@
 #!/bin/sh
 set -e
 
+# we install with the daemon disabled, the configure hook
+# will start it if appropriate
```
Comment (Saviq) [inline snap/hooks/install]:
```suggestion
# will start and enable it if appropriate
```

File: snap/hooks/post-refresh [hunk 2]
```diff
@@ -1,6 +1,10 @@
 #!/bin/sh
 set -e
 
+# we install with the daemon disabled, the configure hook
+# will start it if appropriate
```
Comment (Saviq) [inline snap/hooks/post-refresh]:
```suggestion
# will start and enable it if appropriate
```

File: snap/hooks/configure [hunk 3]
```diff
@@ -151,3 +147,16 @@ if [ -n "${showsplash}" ]; then
   fi
 fi
 
+if [ "$(snapctl get daemon)" = "true" ]; then
+  if snapctl services ${SNAP_NAME}.daemon | grep -q inactive; then
+    snapctl start --enable $SNAP_NAME.daemon 2>&1 || true
+  fi
```
Comment (Saviq) [inline snap/hooks/configure]: Do we need the if? `start --enable…` should be idempotent?

File: snap/hooks/configure [hunk 4]
```diff
@@ -151,3 +147,16 @@ if [ -n "${showsplash}" ]; then
...
+else
+  if snapctl services ${SNAP_NAME}.daemon | grep -q enabled; then
+    snapctl stop --disable $SNAP_NAME.daemon 2>&1 || true
+  fi
```
Comment (Saviq) [inline snap/hooks/configure]: Same here?

File: snap/hooks/configure [hunk 5]
```diff
@@ -151,3 +147,16 @@ if [ -n "${showsplash}" ]; then
...
+fi
+
+if [ $(($config_changes+$display_changes)) -gt 0 ]; then
+  snapctl restart $SNAP_NAME
+fi
```
Comment (Saviq) [inline snap/hooks/configure]: Any way to avoid this if we just started the service above?

**Review by Saviq — CHANGES_REQUESTED** [review]

File: snap/hooks/configure (hunk 4 above)
Comment (Saviq) [reply snap/hooks/configure]:
```suggestion
  snapctl stop --disable $SNAP_NAME.daemon 2>&1 || true
```

File: glue/bin/run-daemon [hunk 6]
```diff
@@ -1,6 +1,4 @@
 #!/bin/sh
 set -e
 
-if [ "$(snapctl get daemon)" = "true" ]; then
-    exec nohup "$@"
-fi
\ No newline at end of file
+nohup "$@"
```
Comment (Saviq) [inline glue/bin/run-daemon]: We didn't need the `exec` in the first place, did we?

**Review by Saviq — APPROVED** [review]
Just one question, otherwise:

> - [x] On installation daemon should default to "true" on Ubuntu Core systems and "false" on "classic" systems;
> - [x] The mir-kiosk.daemon daemon should run if and only if daemon=true; and,
> - [x] (what this changes) snapctrl stop ... should not be called from the daemon launch script, only from hooks.

File: glue/bin/run-daemon (hunk 6 above)
Comment (AlanGriffiths) [reply glue/bin/run-daemon]: Fixed

**Review by Saviq — APPROVED** [review]
Yup.

**Pull request merged**